src/Microsoft.AspNetCore.Blazor.Browser/Interop/TypeInstance.cs

```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Blazor.Browser.Interop
+{
+    internal class TypeInstance
```

Comment:
TBH I'm not super keen on the existence of this type or the way we're representing a type specifier during the call, because:

* It forces us to go through extra JSON deserialization to interpret the incoming type identifier
* It allocates a lot of heap objects on every interop call

Instead of this, what if the JS side continued to accept a type instance object in the way it already does, but before calling .NET, flattens it into a CLR type name string (e.g., `MyGenericType`1[[MyType,MyAssembly]],MyGenericTypeAssembly`). Then all the .NET side has to do is call [`Type.GetType`](https://msdn.microsoft.com/en-us/library/w3f99sx1(v=vs.110).aspx) with that value.

Maybe it's possible to do the equivalent with the method name too, to avoid needing to create and JSON-parse `MethodInstance`s.

What do you think? Does this make this simpler or more complex, and do you think the perf gains would warrant it?

Reply:
**Update** Reading your code below, I realise that's almost exactly what you're already doing. The only difference is that you're doing the type name formatting on the .NET side. I think it would be advantageous to moving that formatting over to the JS side so we can just pass a single flat string.

Reply:
I have a better approach for this, but I need to flush it. I'm planning to break the invocation in two steps.
* Step one calls .net to find the metadata token for a given method info (and caches it).
* Step two invokes the inter-op and passes the metadata token along to find the method. Subsequent calls just use the cached metadata token.

Reply:
What do you mean by a metadata token, and how are you constructing the cache key on the JS side? I'm not sure how this will be advantageous.

If you're constructing something like a string as cache key on the JS side, then you could reduce it to one step anyway, by having the cache key be the CLR representation of a type+method identifier, sending that to the .NET side and having the cache mapping identifier->MethodHandle on the .NET side.

Reply:
Thanks for the explanation about this! I understand more of what you're going for now.

One caveat is that it turns out metadata tokens aren't unique across different .NET modules. So if we are to use metadata tokens, then we'd also need to send the assembly name with each interop call in step two.

You mentioned that one way around this would be to maintain caches on both sides. That makes sense, though if we are doing that then I guess there's no longer any reason to use metadata token rather than `MethodHandle`. On the .NET side we could maintain a `ConcurrentDictionary<int, MethodHandle>` where the keys are just unique integers that our code assigns for each "step one" (method lookup) call it receives. This would make it unnecessary to do the "metadata token to MethodHandle conversion" lookup operation during step 2.

Reply:
@rynowak I think you and I are making the same suggestion. But @javiercn discussed and he explained his reasoning was to capture a method handle in the first phase and then replay it every time the same method is being invoked, so that the total message size is lower for the 2nd and subsequent calls for a given method.

Let us know if you mean something different though!